Sources/MultipartKit/MultipartParser.swift

```
@@ -13,192 +13,314 @@ import CMultipartParser
 ///
 /// Seealso `form-urlencoded` encoding where delimiter boundaries are not required.
 public final class MultipartParser {
-    public var onHeader: (String, String) -> ()
-    public var onBody: (inout ByteBuffer) -> ()
-    public var onPartComplete: () -> ()
+    private enum Error: Swift.Error {
+        case syntax
+    }
 
-    private var callbacks: multipartparser_callbacks
-    private var parser: multipartparser
+    private enum CRLF {
+        case cr, lf
+    }
 
     private enum HeaderState {
-        case ready
-        case field(field: String)
-        case value(field: String, value: String)
+        case preHeaders(CRLF = .cr)
+        case headerName([UInt8] = [])
+        case headerValue([UInt8] = [], name: [UInt8])
+        case postHeaderValue([UInt8], name: [UInt8])
+        case postHeaders
     }
 
-    private var headerState: HeaderState
+    private enum State {
+        case preamble(boundaryMatchIndex: Int = 0)
+        case headers(state: HeaderState = .preHeaders())
+        case body(lowerBound: Int, boundaryMatchIndex: Int = 0)
+        case epilogue
+    }
+
+    public var onHeader: (String, String) -> ()
+    public var onBody: (inout ByteBuffer) -> ()
+    public var onPartComplete: () -> ()
+
+    private let boundary: [UInt8]
+    private let boundaryLength: Int
+    private var state: State
+    private var buffer: ByteBuffer!
+    private var sliceBuffer: ByteBuffer!
 
     /// Creates a new `MultipartParser`.
+    /// - Parameter boundary: boundary separating parts. Must not be empty nor longer than 70 characters according to rfc1341 but we don't check for the latter.
     public init(boundary: String) {
+        precondition(!boundary.isEmpty)
```

Is this safe? If there's a chance we could hit this then we should throw an error instead of bringing down the app